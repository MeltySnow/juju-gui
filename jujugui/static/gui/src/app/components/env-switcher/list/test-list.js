/* Copyright (C) 2017 Canonical Ltd. */
'use strict';

const React = require('react');
const enzyme = require('enzyme');

const EnvList = require('./list');

const jujulibTestHelper = require('@canonical/jujulib/api/test-helpers');
const jujulibModelManager = require('@canonical/jujulib/api/facades/model-manager-v4.js');
const modelResponse = require('@canonical/jujulib/tests/data/modelmanager-response');

describe('EnvList', function() {
  const acl = {canAddModels: sinon.stub().returns(true)};

  let jujuConnection = null;
  let jujuWebsocket = null;

  beforeEach(done => {
    const options = {
      facades: [
        jujulibModelManager
      ]
    };

    const loginResponse = {
      facades: [{name: 'ModelManager', versions: [4]}]
    };

    jujulibTestHelper.makeConnectionWithResponse(
      assert, options, loginResponse, (conn, ws) => {
        jujuConnection = conn;
        jujuWebsocket = ws;
        done();
      });
  });

  const renderComponent = (options = {}) => enzyme.shallow(
    <EnvList
      acl={options.acl || acl}
      changeState={options.changeState || sinon.stub()}
      environmentName={options.environmentName || 'model-name-1'}
      envs={options.envs || []}
      handleModelClick={options.handleModelClick || sinon.stub()}
      switchModel={options.switchModel || sinon.stub()}
      user={options.user || {username: 'who@external', displayName: 'who'}} />
  );

  const getModelsStub = () => {
    jujuWebsocket.queueResponses(new Map([
      [2, modelResponse.listModelSummaries]
    ]));
    return jujuConnection.facades.modelManager.listModelSummaries();
  };

  it('renders a list of models', async function() {
    const models = await getModelsStub();
    const wrapper = renderComponent({envs: models.results});
    expect(wrapper).toMatchSnapshot();
  });

  it('orders the model list, and handles never connected ones', async function() {
    const response = modelResponse.listModelSummaries;
    const results = response.response.results;
    results[0].result['last-connection'] = new Date('July 20, 69 00:20:18 GMT+00:00');
    results[1].result['last-connection'] = new Date('July 20, 69 00:00:18 GMT+00:00');
    results[2].result['last-connection'] = new Date('July 20, 69 00:10:18 GMT+00:00');
    jujuWebsocket.queueResponses(new Map([
      [2, response]
    ]));
    const models = await jujuConnection.facades.modelManager.listModelSummaries();
    const wrapper = renderComponent({envs: models.results});
    expect(wrapper).toMatchSnapshot();
  });

  it('displays only the create new button if there are no models', function() {
    const wrapper = renderComponent();
    assert.strictEqual(wrapper.find('EnvList').length, 0);
  });

  it('clicking a model calls the handleModelClick prop', async function() {
    const models = await getModelsStub();
    const handleModelClick = sinon.stub();
    const getAttribute = sinon.stub();
    getAttribute.withArgs('data-id').returns('abc123');
    getAttribute.withArgs('data-name').returns('the name');
    getAttribute.withArgs('data-owner').returns('who@external');
    const wrapper = renderComponent({envs: models.results, handleModelClick});
    wrapper.find('.env-list__environment').at(0).simulate('click', {
      currentTarget: {
        getAttribute: getAttribute
      }
    });
    assert.equal(handleModelClick.callCount, 1);
  });

  it('new model call is made when clicking on the create model button', async function() {
    const models = await getModelsStub();
    const handleModelClick = sinon.stub();
    const wrapper = renderComponent({envs: models.results, handleModelClick});
    wrapper.find('CreateModelButton').props().action();
    assert.equal(handleModelClick.callCount, 1);
  });

  it('new model is not made when user has incorrect permissions', () => {
    const acl = {canAddModels: sinon.stub().returns(false)};
    const wrapper = renderComponent({acl: acl});
    assert.strictEqual(wrapper.find('CreateModelButton').prop('disabled'), true);
  });
});
