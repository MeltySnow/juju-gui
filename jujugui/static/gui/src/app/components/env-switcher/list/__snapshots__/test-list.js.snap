// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvList orders the model list, and handles never connected ones 1`] = `
<Panel
  focus={true}
  instanceName="env-list-panel"
  visible={true}
>
  <ul
    aria-expanded="true"
    aria-hidden="false"
    aria-labelledby="environmentSwitcherToggle"
    className="env-list"
    id="environmentSwitcherMenu"
    role="menubar"
  >
    <li
      className="env-list__environment"
      data-id="57650e3c-815f-4540-89df-81fd5d70b7ef"
      data-name="group-test"
      key="57650e3c-815f-4540-89df-81fd5d70b7ef"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/group-test
      <div
        className="env-list__last-connected"
      >
        <span>
          Last accessed 
          <DateDisplay
            date="1969-07-20T00:20:18.000Z"
            relative={true}
          />
        </span>
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="9ed7302e-6f22-4c87-8ac5-68bb38efb843"
      data-name="single-cred"
      key="9ed7302e-6f22-4c87-8ac5-68bb38efb843"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/single-cred
      <div
        className="env-list__last-connected"
      >
        <span>
          Last accessed 
          <DateDisplay
            date="1969-07-20T00:10:18.000Z"
            relative={true}
          />
        </span>
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="9ed7302e-6f44-4c87-8ac5-68bb38efb843"
      data-name="mymodel"
      key="9ed7302e-6f44-4c87-8ac5-68bb38efb843"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/mymodel
      <div
        className="env-list__last-connected"
      >
        <span>
          Last accessed 
          <DateDisplay
            date="1969-07-20T00:00:18.000Z"
            relative={true}
          />
        </span>
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="06b11eb5-abd6-48f3-8910-b54cf5905e60"
      data-name="kpi"
      key="06b11eb5-abd6-48f3-8910-b54cf5905e60"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      radman/kpi
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="4e8a5213-c34a-4ac9-83fa-1175165467d7"
      data-name="ci"
      key="4e8a5213-c34a-4ac9-83fa-1175165467d7"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      bigboss/ci
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
  </ul>
  <CreateModelButton
    action={[Function]}
    changeState={[Function]}
    disabled={false}
    switchModel={[Function]}
    title="Start a new model"
    type="neutral"
  />
</Panel>
`;

exports[`EnvList renders a list of models 1`] = `
<Panel
  focus={true}
  instanceName="env-list-panel"
  visible={true}
>
  <ul
    aria-expanded="true"
    aria-hidden="false"
    aria-labelledby="environmentSwitcherToggle"
    className="env-list"
    id="environmentSwitcherMenu"
    role="menubar"
  >
    <li
      className="env-list__environment"
      data-id="57650e3c-815f-4540-89df-81fd5d70b7ef"
      data-name="group-test"
      key="57650e3c-815f-4540-89df-81fd5d70b7ef"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/group-test
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="9ed7302e-6f44-4c87-8ac5-68bb38efb843"
      data-name="mymodel"
      key="9ed7302e-6f44-4c87-8ac5-68bb38efb843"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/mymodel
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="9ed7302e-6f22-4c87-8ac5-68bb38efb843"
      data-name="single-cred"
      key="9ed7302e-6f22-4c87-8ac5-68bb38efb843"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      thedr/single-cred
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="06b11eb5-abd6-48f3-8910-b54cf5905e60"
      data-name="kpi"
      key="06b11eb5-abd6-48f3-8910-b54cf5905e60"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      radman/kpi
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
    <li
      className="env-list__environment"
      data-id="4e8a5213-c34a-4ac9-83fa-1175165467d7"
      data-name="ci"
      key="4e8a5213-c34a-4ac9-83fa-1175165467d7"
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
    >
      bigboss/ci
      <div
        className="env-list__last-connected"
      >
        Unknown last connection
      </div>
    </li>
  </ul>
  <CreateModelButton
    action={[Function]}
    changeState={[Function]}
    disabled={false}
    switchModel={[Function]}
    title="Start a new model"
    type="neutral"
  />
</Panel>
`;
